tests/gradients/test_finite_difference.py

```diff
@@ -0,0 +1,590 @@
+            qml.expval(qml.PauliX(1))
+
+        tapes, fn = finite_diff(tape, order=order, form=form)
+        res = fn(dev.batch_execute(tapes))
+        assert res.shape == (2, 2)
+
+        expected = np.array([[-np.sin(x), 0], [0, np.cos(y)]])
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+    def test_var_expectation_values(self, order, form, tol):
+        """Tests correct output shape and evaluation for a tape
+        with expval and var outputs"""
+        dev = qml.device("default.qubit", wires=2)
+        x = 0.543
+        y = -0.654
+
+        with qml.tape.JacobianTape() as tape:
+            qml.RX(x, wires=[0])
+            qml.RY(y, wires=[1])
+            qml.CNOT(wires=[0, 1])
+            qml.expval(qml.PauliZ(0))
+            qml.var(qml.PauliX(1))
+
+        tapes, fn = finite_diff(tape, order=order, form=form)
+        res = fn(dev.batch_execute(tapes))
+        assert res.shape == (2, 2)
+
+        expected = np.array([[-np.sin(x), 0], [0, -2 * np.cos(y) * np.sin(y)]])
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+    def test_prob_expectation_values(self, order, form, tol):
+        """Tests correct output shape and evaluation for a tape
+        with prob and expval outputs"""
+        dev = qml.device("default.qubit", wires=2)
+        x = 0.543
+        y = -0.654
+
+        with qml.tape.JacobianTape() as tape:
+            qml.RX(x, wires=[0])
+            qml.RY(y, wires=[1])
+            qml.CNOT(wires=[0, 1])
+            qml.expval(qml.PauliZ(0))
+            qml.probs(wires=[0, 1])
+
+        tapes, fn = finite_diff(tape, order=order, form=form)
+        res = fn(dev.batch_execute(tapes))
+        assert res.shape == (5, 2)
+
+        expected = (
+            np.array(
+                [
+                    [-2 * np.sin(x), 0],
+                    [
+                        -(np.cos(y / 2) ** 2 * np.sin(x)),
+                        -(np.cos(x / 2) ** 2 * np.sin(y)),
+                    ],
+                    [
+                        -(np.sin(x) * np.sin(y / 2) ** 2),
+                        (np.cos(x / 2) ** 2 * np.sin(y)),
+                    ],
+                    [
+                        (np.sin(x) * np.sin(y / 2) ** 2),
+                        (np.sin(x / 2) ** 2 * np.sin(y)),
+                    ],
+                    [
+                        (np.cos(y / 2) ** 2 * np.sin(x)),
+                        -(np.sin(x / 2) ** 2 * np.sin(y)),
+                    ],
+                ]
+            )
+            / 2
+        )
+
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+
+@pytest.mark.parametrize("order", [2])
+@pytest.mark.parametrize("form", ["center"])
+class TestFiniteDiffGradients:
+    """Test that the transform is differentiable"""
+
+    def test_autograd(self, order, form, tol):
+        """Tests that the output of the finite-difference transform
+        can be differentiated using autograd, yielding second derivatives."""

```

Comment:
I'm a bit confused about where the second derivative comes in since `n=1` in the transform. Is it because the base tape is a `JacobianTape`?

Reply:
Nope, it's because the device is `default.qubit.autograd`!

1. I am writing a cost function that computes the finite-difference gradient of this QNode using this transform.
2. Since this transform is differentiable, I can differentiate the cost function
3. When the autodiff framework performs backpropagation, it will end up needing to compute various `d tape/d parameters` as part of the chain rule. However, since the device is written using autograd, more chain rules are applied to simply get to `d state/d parameters`!

So it's a mixture of (a) the transform being differentiable, and (b) the device execution being differentiable.

Reply:
Basically, you can differentiate any `qfunc(qnode)` cost function, to get the gradient `d qfunc(qnode)/d parameters`. In this case, since `qfunc` itself is the first derivative, we end up with the second derivative 😆